src/snowflake/snowpark/mock/_functions.py

```diff
@@ -413,53 +413,47 @@ def mock_to_decimal(
 
     [x] For fractional input, the precision is deduced as the number of digits after the point.
 
     For VARIANT input:
 
-        [ ] If the variant contains a fixed-point or a floating-point numeric value, an appropriate numeric conversion is performed.
+        [x] If the variant contains a fixed-point or a floating-point numeric value, an appropriate numeric conversion is performed.
 
-        [ ] If the variant contains a string, a string conversion is performed.
+        [x] If the variant contains a string, a string conversion is performed.
 
-        [ ] If the variant contains a Boolean value, the result is 0 or 1 (for false and true, correspondingly).
+        [x] If the variant contains a Boolean value, the result is 0 or 1 (for false and true, correspondingly).
 
-        [ ] If the variant contains JSON null value, the output is NULL.
+        [x] If the variant contains JSON null value, the output is NULL.
     """
-    res = []
-
-    for data in e:
-        if data is None:
-            res.append(data)
-            continue
-        try:
-            try:
-                float(data)
-            except ValueError:
-                raise SnowparkSQLException(f"Numeric value '{data}' is not recognized.")
 
-            integer_part = round(float(data))
-            integer_part_str = str(integer_part)
-            len_integer_part = (
-                len(integer_part_str) - 1
-                if integer_part_str[0] == "-"
-                else len(integer_part_str)
+    def cast_as_float_convert_to_decimal(x: Union[Decimal, float, str, bool]):
+        x = float(x)
+        if x in (math.inf, -math.inf, math.nan):
+            raise ValueError(
+                "Values of infinity and NaN cannot be converted to decimal"
             )
-            if len_integer_part > precision:
-                raise SnowparkSQLException(f"Numeric value '{data}' is out of range")
-            remaining_decimal_len = min(precision - len(str(integer_part)), scale)
-            res.append(Decimal(str(round(float(data), remaining_decimal_len))))
-        except BaseException:
-            if try_cast:
-                res.append(None)
-            else:
-                raise
-
-    return ColumnEmulator(
-        data=res,
-        sf_type=ColumnType(DecimalType(precision, scale), nullable=e.sf_type.nullable),
+        integer_part = str(round(abs(x)))
+        if len(integer_part) > precision:
+            raise SnowparkSQLException(f"Numeric value '{x}' is out of range")
+        remaining_decimal_len = min(precision - len(integer_part), scale)
+        return Decimal(str(round(x, remaining_decimal_len)))
+
+    if isinstance(e.sf_type.datatype, (_NumericType, BooleanType, NullType)):
+        res = e.apply(
+            lambda x: try_convert(cast_as_float_convert_to_decimal, try_cast, x)
+        )
+    elif isinstance(e.sf_type.datatype, (StringType, VariantType)):
+        res = e.replace({"E": 0}).apply(
```

Comment:
This is because `For floating-point input, omitting the mantissa or exponent is allowed and is interpreted as 0. Thus, E is parsed as 0.`. Normally, something like `3.14E-10` would just be casted by `float()`, but `E` is an edge case that works in Snowflake but not in Python.